bazaarvoice/emodb
# Added global settings for configurable suppressed databus events

## Github Issue #

[26](https://github.com/bazaarvoice/emodb/issues/26)

## What Are We Doing Here?

Three changes:

* The condition for ignorable suppressed databus events is now configurable rather than hard-coded to `not({..,"~tags",contains("re-etl")})`.
* A new DW task was created for updating the condition.
* A re-usable settings-management system was created for storing and propagating the condition through the system.

## How to Test and Verify

1. Create a databus subscription:

   ```
   $ curl -s -XPUT -H "Content-Type: application/x.json-condition" 'http://localhost:8080/bus/1/test-sub?ignoreSuppressedEvents=true' --data-binary 'alwaysTrue()'
   {"success":true}
   ```
2. Assuming a clean system there should be not suppressable events configured:

   ```
   $ curl -s 'http://localhost:8080/bus/1/test-sub'
   {"name":"test-sub","tableFilter":"alwaysTrue()","expiresAt":"2016-09-03T15:47:51.146Z","eventTtl":86400000}
   ```
3. Change the global condition:

   ```
   $ curl -XPOST 'localhost:8081/tasks/databus-suppressed-event-condition?value=\{..,"~tags":containsAny("ignore")\}'
   Prior value:   alwaysFalse()
   Updated value: {..,"~tags":contains("ignore")}
   ```
4. Renew the subscription; it should now contain the suppressed condition:

   ```
   $ curl -s -XPUT -H "Content-Type: application/x.json-condition" 'http://localhost:8080/bus/1/test-sub?ignoreSuppressedEvents=true' --data-binary 'alwaysTrue()'
   {"success":true}
   $curl -s 'http://localhost:8080/bus/1/test-sub'
   {"name":"test-sub","tableFilter":"and(alwaysTrue(),not({..,\"~tags\":contains(\"ignore\")}))","expiresAt":"2016-09-03T15:53:38.180Z","eventTtl":86400000}
    ```

## Risk

Fairly low; the area of the system affected is very localized, and the more complicated code around inserting the suppressed event condition was already in place.

### Level 

Medium

### Required Testing

Manual

### Risk Summary

This change doesn't really add many new risks.  The largest is the uncertainty due to natural delays.  Cross-data-center replication of the setting may take up to a minute, longer if there is a break in the
Cassandra replication.  Additionally, the ignored events are only updated per-subscription at renew time.  So, for example, if a subscription only renews once every four hours then updating the suppressed event condition immediately may take up to four hours to go into effect for that subscription.  Not really a bug, just something to keep in mind when changing the condition; the administrator probably wants to make the update with ample lead time before anyone floods the system with events that should be suppressed.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
